### PR TITLE
[BEAM-6418] Execute Flink tests serially to avoid memory issues

### PR DESCRIPTION
**Pull request by mxm**
We had previously disabled Flink tests for 1.6 and 1.7 due to memory issues when
they executed in parallel. This lets them execute one after another on Jenkins.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by mxm** [comment]
The serial solution makes sense, should this be a harder to fix problem. It is also inconvenient that at the moment the tests can't be executed at all for 1.6/1.7 without manual intervention.

Attempting to fix this here: https://github.com/apache/beam/pull/7512

**Comment by mxm** [comment]
@adude3141 Could you have a look?

File: runners/flink/flink_runner.gradle [hunk 1]
```diff
@@ -68,10 +68,15 @@ test {
   if (System.getProperty("beamSurefireArgline")) {
     jvmArgs System.getProperty("beamSurefireArgline")
   }
-}.onlyIf {
-  // TODO Running tests of all Flink versions in parallel can be too harsh on Jenkins memory
-  // Skip 1.6/1.7 tests for now, to avoid "Exit code 137", i.e. Jenkins host killing the Gradle test process
-  project.name != "beam-runners-flink-1.6" && project.name != "beam-runners-flink-1.7"
+  if (project.jenkins.isCIBuild) {
```
Comment (adude3141) [inline runners/flink/flink_runner.gradle]: Wont we have similar issues on dev machine? Or do we expect those to have more resources available? From my point of view, we could enable that serialisation also on 'non ci builds'?

File: runners/flink/flink_runner.gradle [hunk 2]
```diff
@@ -68,10 +68,15 @@ test {
...
+    // TODO Running tests of all Flink versions in parallel can be too harsh on Jenkins memory
+    // Run them serially for now, to avoid "Exit code 137", i.e. Jenkins host killing the Gradle test process
+    if (project.name == "beam-runners-flink-1.6") {
+      dependsOn(":beam-runners-flink_2.11:test")
```
Comment (adude3141) [inline runners/flink/flink_runner.gradle]: Should we prefer `mustRunAfter`here? `dependsOn`will add a dependency, so if we intend to execute 'beam-runners-flink-1.6' that 'beam-runners-flink_2.11:test' will also be run, i.e. gradle will add the latter to the execution graph even if it was not requested.

In contrast, 'mustRunAfter' will only order the task, if *both* are present on the execution graph. If only the former is requested, the latter will not run.

File: runners/flink/flink_runner.gradle [hunk 3]
```diff
@@ -68,10 +68,15 @@ test {
...
+    } else if (project.name == "beam-runners-flink-1.7") {
+      dependsOn(":beam-runners-flink-1.6:test")
```
Comment (adude3141) [inline runners/flink/flink_runner.gradle]: same here?

File: runners/flink/flink_runner.gradle (hunk 2 above)
Comment (adude3141) [reply runners/flink/flink_runner.gradle]: although this will require to specify the total order, i.e. declare all combinations

**Comment by adude3141** [comment]
Overall I am asking myself why is this only a 'flink-internal' problem, i.e. don't we have similar tests with e.g. spark? And wouldn't those tests also cause this issue? If so, these dependsOn and mustRunAfter will probably not scale.

Unsure, whether the even hackier 'solution' of writing to a common outfolder would ease that...

**Comment by mxm** [comment]
I was asking myself the same question. I've tried lowering memory usage in tests, but that did not fix it in all occasions. No other Runners execute three different full-blown testing clusters at the same time. I've never had this problem with parallel execution locally.

However, I've seen the "Exit Code 137" (killed by OS) also in other memory intensive tests on Jenkins. I'm assuming at the moment it is not Flink specific, but a very memory constraint test environment.

File: runners/flink/flink_runner.gradle (hunk 1 above)
Comment (mxm) [reply runners/flink/flink_runner.gradle]: I've never had these problems locally. I've turned this off for local execution because it is much slower when you only work on the flink modules.

File: runners/flink/flink_runner.gradle (hunk 2 above)
Comment (mxm) [reply runners/flink/flink_runner.gradle]: Yes, that was my intention and I couldn't find another way of doing this. Thank you will update the PR.

File: runners/flink/flink_runner.gradle (hunk 1 above)
Comment (mxm) [reply runners/flink/flink_runner.gradle]: I've removed the Jenkins switch, to make it consistent. Now that we use mostRunAfter, it behaves much nicer on the dev machine.

**Comment by adude3141** [comment]
LGTM as it will enable those currently disabled tests again.

Of course we need to find a more scalable and permanent solution for those 'killed by OS' failures...

Nit: The current ordering is incomplete in the sense, that if we run only ':beam-runners-flink_2.11:test' and ':beam-runners-flink-1.7:test' they still might be executed in parallel (as we implemented no order here). But it seems ok to me to ignore that corner case.

**Comment by mxm** [comment]
Thanks. Good catch, the `mustRunAfter" semantics are different from `dependsOn`. We must spell out all the possibilities.

**Comment by adude3141** [comment]
Unfortunately yes. Which makes it 'non-scalable' across multi project build.

If this will ever be a requirement, we might try something along the lines as adding a function `runSequentially` (or some such) anchored at some root plugin which will collect all those tasks and do that 'must run after' in more automated fashion...

Something to be used simply as
```
test {
  // configure as usual
  ...
  // add to list of sequential test
  runSequentially() 
}
```
dunno if that works, though. and of course rather ugly ...

**Comment by mxm** [comment]
Flaky unrelated test in Java PreCommit: https://builds.apache.org/job/beam_PreCommit_Java_Commit/4239/

**Comment by mxm** [comment]
Run Java PreCommit
